### DISCUSSION TITLE
S3 File Streaming

### DISCUSSION THREAD

**Discussion by apikee** [pocketbase/pocketbase#1976] [Q&A]
Hello, i would like to use S3 storage with my PB instance for media streaming, but I would like to know following:
In Apple Podcast requirements is stated this: https://podcasters.apple.com/support/823-podcast-requirements
```
Display your show artwork and allow streaming playback of your episodes by enabling your hosting server for HTTP HEAD requests and byte-range requests.
```
S3 by itself supports byte-range requests (https://docs.aws.amazon.com/whitepapers/latest/s3-optimizing-performance-best-practices/use-byte-range-fetches.html) but will this be compatible through PB?
Thanks

**Comment by ganigeorgiev** [comment]
Partial/byte-range requests should be supported since v0.9+ (see https://github.com/pocketbase/pocketbase/issues/1125) so if you are streaming directly with PocketBase (aka. without a reverse proxy) it should work without any extra configuration.

**Reply by ganigeorgiev** [reply]
> ...enabling your hosting server for HTTP HEAD requests...

This is the only part that we currently don't have out of the box.
The `/api/files/:collection/:recordId/:filename` endpoint currently works only with `GET` requests.
I'll consider adding a `HEAD` requests support for the next v0.14/v0.15 release.

**Reply by apikee** [reply]
Great, thanks for the reply. Is there some "easy" way of implementing the HEAD endpoint myself? I was thinking to create custom "proxy" endpoint which would accept both byte-range headers and the HEAD request -> this endpoint would take the byte-range headers and forward them to the S3 file url and then return the ranged response. But it's just a speculation, i'm not really familiar with the S3 storage. I presume this would work, but not 100% sure. But if it would be supported out of the box without the need to make custom endpoint, it would be great. Btw. why the reverse proxy is an issue? Do you think it would be an issue even with the implementation i mentioned?

**Reply by ganigeorgiev** [reply]
>  Is there some "easy" way of implementing the HEAD endpoint myself?  I was thinking to create custom "proxy" endpoint which would accept both byte-range headers and the HEAD request -> this endpoint would take the byte-range headers and forward them to the S3 file url and then return the ranged response.

Technically, it should be enough to add the same [GET file download route](https://github.com/pocketbase/pocketbase/blob/master/apis/file.go#L21) but with a `HEAD`.
This should work because internally the actual file serving is done by the standard `http.ServeContent`, which'll [check for the request method and skip the body write](https://cs.opensource.google/go/go/+/master:src/net/http/fs.go;l=352?q=http.ServeContent&ss=go%2Fgo). But it needs to be tested.

> Btw. why the reverse proxy is an issue? Do you think it would be an issue even with the implementation i mentioned?

It is not an issue, it could actually be a valid workaround. You can implement the proxy cache workaround if you don't want to wait for the next release (_there are no ETAs yet_) mentioned in https://github.com/pocketbase/pocketbase/issues/1125#issuecomment-1328036972.

**Reply by apikee** [reply]
I think I can wait for the release, I see you release new versions quite often

**Reply by apikee** [reply]
@ganigeorgiev this is exactly what I had in mind: https://medium.com/@radhian.amri/video-streaming-using-http-206-partial-content-in-go-4e89d96abdd0
Echo is obviously making it easy

**Reply by ganigeorgiev** [reply]
@apikee
The range requests handling comes from the standard library - `http.ServeContent`.
It should work with any `io.ReadSeaker` compatible struct (including the `blob.Reader` from the [google/go-cloud/blob](https://github.com/google/go-cloud/tree/master/blob) that we currently use).

Since the lack of `HEAD` request could be a blocker for users who rely on previewing media files in Safari, I've pushed a fix as minor [v0.13.1 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.13.1).

Please also note that for better performance when serving large files with enabled S3 storage, you may want to use directly the S3 CDN url (if available) instead of the `/api/files/...` route (_there will be eventually option for this in the Admin UI as part of #28 and its PR #986_). The raw S3 object name follows the format: `/collectionId/recordId/filename`.

**Reply by apikee** [reply]
wow, thank you for the quick action!
